### PR TITLE
API GW database and actions deployment

### PR DESCRIPTION
**Pull request by mdeuser**
API GW database creation and wipe
API GW backend route management action and package creation
backend route management action negative parameter tests

### CODE REVIEW

**Comment by rabbah** [comment]
a quick comment following a cursory scan - there is quite a bit of clone and own logic for accessing the db, did you consider pulling that out into a single file and requiring it in the resultant smaller "actions" and use zip files to package the dependencies?

File: ansible/roles/routemgmt/files/gwapis_views.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "_id": "_design/gwapis",
+  "views": {
+    "routes-by-path": {
+      "map": "function (doc) {\n  var PATHSEP = \"/\";\n\n  if (doc.namespace && doc.gatewayPath) {\n    emit([doc.namespace, doc.gatewayPath], {\"gatewayMethod\":doc.gatewayMethod, \"gatewayPath\":doc.gatewayPath, \"action\":doc.action, \"route-collection\":doc[\"route-collection\"], \"apidoc\":doc.apidoc});\n  }\n}"
```
Comment (markusthoemmes) [inline ansible/roles/routemgmt/files/gwapis_views.json]: Will there be any documents in the database that are not routes?

Comment (mdeuser) [inline ansible/roles/routemgmt/files/gwapis_views.json]: @markusthoemmes - at the moment, no.  But for performance, I can imagine views that return a subset of data from the docs

**Comment by mdeuser** [comment]
@rabbah - Good point.  Yes, the packaging is not currently optimal.  With the node packaging support (.zip) I think this can be repackaged a bit more cleanly.  I'll open an issue for that

**Comment by mdeuser** [comment]
PG1 578 (API GW tests all passed)

**Comment by mdeuser** [comment]
@markusthoemmes - Do you have time to complete the review this PR?  I'll squash/rebase when all looks good.

One question I have is how to resolve the following actor failures that occur during test (and not just in this PR's new tests).  Any ideas?
```
[2016-11-15T21:00:49.244Z] [ERROR] Outgoing request stream error [akka.actor.ActorSystemImpl:71] 
    akka.stream.AbruptTerminationException: Processor actor [Actor[akka://default/user/StreamSupervisor-521/flow-5-0-unknown-operation#-363075754]] terminated abruptly
```

File: tools/db/wipeTransientDBs.sh [hunk 2]
```diff
@@ -55,6 +55,20 @@ do
...
+# drop the database
+CMD="$CURL_ADMIN -X DELETE $URL_BASE/$DB_APIGW"
+RES=$($CMD)
+if [[ "$RES" == '{"ok":true}' ]]; then
+    echo DELETED
+else
+    # table may not exist
+    echo WARNING: $RES
+fi
+
```
Comment (markusthoemmes) [inline tools/db/wipeTransientDBs.sh]: I feel like we should not conflate the standard whisks db with the API gateway db. Any reason we need to delete the DB in this script?

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+              Seq("-p", "apidoc", "{\"namespace\":\"_\",\"gatewayBasePath\":\"/ApiGwRoutemgmtActionTests_bp\",\"gatewayPath\":\"ApiGwRoutemgmtActionTests_rp\",\"gatewayMethod\":\"get\",\"action\":{\"backendMethod\":\"post\",\"backendUrl\":\"URL\",\"namespace\":\"_\"}}")),
+            ("/whisk.system/routemgmt/createRoute", ANY_ERROR_EXIT, "action is missing the authkey field",
+              Seq("-p", "apidoc", "{\"namespace\":\"_\",\"gatewayBasePath\":\"/ApiGwRoutemgmtActionTests_bp\",\"gatewayPath\":\"ApiGwRoutemgmtActionTests_rp\",\"gatewayMethod\":\"get\",\"action\":{\"backendMethod\":\"post\",\"backendUrl\":\"URL\",\"namespace\":\"_\",\"name\":\"N\"}}")),
+            ("/whisk.system/routemgmt/createRoute", ANY_ERROR_EXIT, "swagger and gatewayBasePath are mutually exclusive and cannot be specified together",
+              Seq("-p", "apidoc", "{\"namespace\":\"_\",\"gatewayBasePath\":\"/ApiGwRoutemgmtActionTests_bp\",\"gatewayPath\":\"ApiGwRoutemgmtActionTests_rp\",\"gatewayMethod\":\"get\",\"action\":{\"backendMethod\":\"post\",\"backendUrl\":\"URL\",\"namespace\":\"_\",\"name\":\"N\",\"authkey\":\"XXXX\"},\"swagger\":{}}")),
```
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Use triple-quote notation to surpass the need to escape the JSON strings.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+        invalidArgs foreach {
+            case (action:String, exitcode:Int, errmsg:String, params:Seq[String]) =>
+                var cmd: Seq[String] = Seq("action",
+                    "invoke",
+                    action,
+                    "-i", "-b", "-r",
+                    "--apihost", wskprops.apihost,
+                    "--auth", whisksystemAuthId+":"+whisksystemAuthPwd
+                ) ++ params
```
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Why write your own version of invoke? Is the usual `wsk.action.invoke` not enough?

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        authstore.shutdown()
+        super.afterAll()
+    }
+
+    behavior of "API Gateway routemgmt action parameter validation"
+
+    it should "reject routemgmt actions that are invoked with not enough parameters" in {
+        val invalidArgs = Seq(
+            //getApi
+            ("/whisk.system/routemgmt/getApi", ANY_ERROR_EXIT, "namespace is required", Seq()),
```
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: All of them use `ANY_ERROR_EXIT`, is that even needed then?

File: ansible/roles/routemgmt/tasks/db.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    method: POST
+    status_code: 200,201
+    body_format: json
+    body: "{{ gwapis_views }}"
+    user: "{{ db_username }}"
+    password: "{{ db_password }}"
+    force_basic_auth: yes
+  when: result.status == 404
+
+- name: create the API Gateway index (Cloudant only)
```
Comment (markusthoemmes) [inline ansible/roles/routemgmt/tasks/db.yml]: Why is this needed?

File: core/routemgmt/activateApi.js [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+  // 2. Create the request document to send to the API Gateway
+  // 3. Send the request to the API Gateway to configure and activate the API
+  // 4. If successful, update the API configuration in the DB to indicate the activation status
+  var retApi;
+  var tenantGuid;
+  var tenantAdded = false;
+  var gwApiActivated = false;
+  var dbUpdated = false;
+  var gwApiDoc;
+  var dbApiDoc;
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: I understand why these definitions are here, but this feels like quite brittle to me as you mutate all those variables in Promises down the road. Is there a way to overcome this `undefined` hell?

File: core/routemgmt/activateApi.js [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+  console.log('GW URL     : '+message.gwUrl);
+  console.log('GW User    : '+message.gwUser);
+  console.log('GW Pwd     : '+message.gwPwd);
+  console.log('GW Auth    : '+gwInfo.gwAuth);
+  console.log('namespace  : '+message.namespace);
+  console.log('basepath   : '+message.basepath);
+  console.log('docid      : '+message.docid);
+
+  var cloudantOrError = getCloudantAccount(message);
+  if (typeof cloudantOrError !== 'object') {
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: Why not return `false` if something went wrong in `getCloudantAccount` instead? Or `null`.

File: core/routemgmt/activateApi.js [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+  console.log('getApiDoc() for: ', params);
+    return whisk.invoke({
+      name: actionName,
+      blocking: true,
+      parameters: params
+    })
+    .then(function (activation) {
+      console.log('whisk.invoke('+actionName+', '+params.namespace+', '+params.basepath+') ok');
+      console.log('Results: '+JSON.stringify(activation));
+    if (activation && activation.result && activation.result.apis &&
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: check the identation

File: core/routemgmt/activateApi.js [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+    .catch(function (error) {
+      console.error('whisk.invoke('+actionName+', '+params.namespace+', '+params.basepath+') error: '+JSON.stringify(error));
+      return Promise.reject(error);
+    });
+}
+
+function addTenantToGateway(gwInfo, namespace) {
+  var options = {
+    url: gwInfo.gwUrl+'/tenants',
+    agentOptions: {rejectUnauthorized: false},
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: This effectively deactivates SSL? Is that what's wanted? Sounds like it should not be in production?

File: core/routemgmt/activateApi.js [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    headers: {
+      'Content-Type': 'application/json',
+      'Accept': 'application/json'
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: Is this necessary when you already set the `json` attribute?

File: core/routemgmt/activateApi.js [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+        if (body && body.id) {  // body has format like:  { id: GUID, namespace: NAMESPACE, instance: 'openwhisk' }
+          console.log('addTenantToGateway: got a single tenant resposne');
+          resolve(body);
+        } else {
+          console.error('addTenantToGateway: failure: No tenant guid provided')
+          reject('Unable to configure the API Gateway: Invalid response from API Gateway');
+        }
+      }
+    });
+  });
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: Should we rely on #1494 for this? Small change, big impact on readability.

File: core/routemgmt/deactivateApi.js [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+  console.log('DB port    : '+message.port);
+  console.log('DB protocol: '+message.protocol);
+  console.log('DB username: '+message.username);
+  console.log('DB database: '+message.dbname);
+  console.log('GW URL     : '+message.gwUrl);
+  console.log('GW Auth API: '+message.gwAuth);
+  console.log('namespace  : '+message.namespace);
+  console.log('basepath   : '+message.basepath);
+
+  var cloudantOrError = getCloudantAccount(message);
```
Comment (markusthoemmes) [inline core/routemgmt/deactivateApi.js]: Is this really just possible on Cloudant? Not CouchDB?

File: core/routemgmt/deleteApi.js [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+    name: actionName,
+    blocking: true,
+    parameters: params
+  })
+  .then(function (activation) {
+    console.log('whisk.invoke('+actionName+', '+params.namespace+', '+params.basepath+') ok');
+    console.log('Results: '+JSON.stringify(activation));
+    if (activation && activation.result && activation.result.apis &&
+        activation.result.apis.length == 1 && activation.result.apis[0].value &&
+        activation.result.apis[0].value._rev) {
```
Comment (markusthoemmes) [inline core/routemgmt/deleteApi.js]: I find that defensive coding quite annoying to read, while i realize its necessary.

Will open an issue to add lodash to our runtimes. You could use https://lodash.com/docs/4.16.4#has then, which would make that way more readable.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 3 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Nice tip.  Wish I'd known that before typing all of those escapes :-)

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 5 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: I included the exit code just in case we need more granular exit code validation.

File: core/routemgmt/deactivateApi.js (hunk 13 above)
Comment (mdeuser) [inline core/routemgmt/deactivateApi.js]: In theory, it should work with CouchDB as well; although, I've not tried that out yet.  Issue #1515

File: core/routemgmt/activateApi.js (hunk 12 above)
Comment (mdeuser) [inline core/routemgmt/activateApi.js]: I agree.  Use of the request-promise package would add a bit more readability.  Alas, that issue is not currently resolved; but I'm thinking that it should not block the initial API GW code delivery.

File: ansible/roles/routemgmt/defaults/main.yml [hunk 15]
```diff
@@ -0,0 +1,3 @@
+apigw_auth_user: ""
+apigw_auth_pwd: ""
+apigw_host: "http://{{ groups['edge']|first }}:3100/v1"
```
Comment (mdeuser) [inline ansible/roles/routemgmt/defaults/main.yml]: This defaults file can now be deleted.  Based on http://docs.ansible.com/ansible/playbooks_variables.html#id31 it seemed to provide default values that could be overridden by other playbook variable settings.  But as builds progressed, these default values were not as pervasive as I had assumed.

File: core/routemgmt/activateApi.js (hunk 8 above)
Comment (mdeuser) [inline core/routemgmt/activateApi.js]: The function is actually reused from the Cloudant package.  I left the interface as-is thinking this function might be leveraged directly from the package when the entire set of API GW actions were re-factored to use single-source common functions..

File: core/routemgmt/activateApi.js [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+  // Log parameter values
+  console.log('DB host    : '+message.host);
+  console.log('DB port    : '+message.port);
+  console.log('DB protocol: '+message.protocol);
+  console.log('DB username: '+message.username);
+  console.log('DB database: '+message.dbname);
+  console.log('GW URL     : '+message.gwUrl);
+  console.log('GW User    : '+message.gwUser);
+  console.log('GW Pwd     : '+message.gwPwd);
+  console.log('GW Auth    : '+gwInfo.gwAuth);
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: Logging credentials seems like a bad idea in a production environment?

File: core/routemgmt/activateApi.js (hunk 7 above)
Comment (mdeuser) [inline core/routemgmt/activateApi.js]: I struggled in several places with the need to use promises, but also with the need to maintain some state across promise processing functions.  At the time I had implemented something even uglier and @bjustin-ibm suggested the using variables that were scoped outside of the promise chain coupled with a single catch.

The retApi variable can be removed; but that does not address your overall point.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+                    "invoke",
+                    action,
+                    "-i", "-b", "-r",
+                    "--apihost", wskprops.apihost,
+                    "--auth", whisksystemAuthId+":"+whisksystemAuthPwd
+                ) ++ params
+                val rr = wsk.cli(cmd, expectedExitCode = exitcode)
+                rr.stderr should include regex (errmsg)
+        }
+    }
```
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Any tests to verify that the actions actually do the right job?

File: core/routemgmt/activateApi.js (hunk 7 above)
Comment (bjustin-ibm) [inline core/routemgmt/activateApi.js]: A couple thoughts here... 

Some of these variables are not actually read, and can be removed.

It may be possible to handle failures more granularly by adding error callbacks as the second argument to `then()`. This will let you handle specific errors as they happen, rather than trying to handle all errors in a single `catch()` at the end.

If you are incrementally building up a result, perhaps it is reasonable to define a single `result = {};` object which is modified as asynchronous steps complete. @markusthoemmes would this address your concern about `undefined` variables?

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 17 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Yes, those will be forthcoming when the API GW is delivered.

File: core/routemgmt/activateApi.js (hunk 10 above)
Comment (mdeuser) [inline core/routemgmt/activateApi.js]: Good observation.  I just checked with the API GW folks and they now use a CA signed cert, so this limitation can be removed.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 4 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: I believe the `wsk.action.invoke` hardwired the use of the whisk.properties auth key (auth.guest).  When invoking these actions, I needed to use the /whisk.system's auth key.

**Comment by mdeuser** [comment]
PG2 599 - passed

**Comment by mdeuser** [comment]
@markusthoemmes - now that a local API GW has been merged, I've added some positive test cases to add coverage for actual interaction with the API GW.

**Comment by mdeuser** [comment]
The recent travis build runs the API GW tests (positive and negative) against a local API GW container running on travis.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala [hunk 18]
```diff
@@ -0,0 +1,470 @@
...
+               swagger: String = "",
+               expectedExitCode: Int = SUCCESS_EXIT): RunResult ={
+        val parms = Map[String,JsValue]() ++
+            Map("namespace" -> "_".toJson) ++
+            { if (basepath.length() > 0)  Map("basepath" -> basepath.toJson) else Map[String,JsValue]() } ++
+            { if (relpath.length() > 0)   Map("relpath" -> relpath.toJson) else Map[String,JsValue]() } ++
+            { if (operation.length() > 0) Map("operation" -> operation.toJson) else Map[String,JsValue]() } ++
+            { if (apiname.length() > 0) Map("apiname" -> apiname.toJson) else Map[String,JsValue]() } ++
+            { if (action != null) Map("action" -> action.toJson) else Map[String,JsValue]() } ++
+            { if (swagger.length() > 0) Map("swagger" -> swagger.toJson) else Map[String,JsValue]() }
```
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Seems like you want all of those to be `Option[String]` instead of checking the length of the String?

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala [hunk 19]
```diff
@@ -0,0 +1,470 @@
...
+            expectedExitCode = SUCCESS_EXIT)(wskprops)
+        var apiJsArray: JsArray =
+            try {
+                var apisobj = rr.stdout.parseJson.asJsObject.fields("apis")
+                apisobj.convertTo[JsArray]
+            } catch {
+                case e:Exception =>
+                    JsArray.empty
+            }
+        return apiJsArray
```
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: One does not simply use `return` in Scala :).

Above code is equivalent to the more idiomatic: `Try(rr.stdout.parseJson.asJsObject.fields("apis").convertTo[JsArray]).getOrElse(JsArray.empty)`

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 4 above)
Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Using the standard  `wsk.action.invoke` you can also greatly simplify the parameter declaration in your testdata.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 19 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: I have played a bit with Scala's implicit `return` feature, but as a personal preference I tend to lean toward an explicit coding style, so I will usually retain the `return` for clarity - and to keep me from accidentally changing return behavior :-)

Re: use of `JsArray`.  Yes, this method could have returned `Vector[JsValue]`, which is one step closer to the desired results.

Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Does the `JsValue` follow a specific format?

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 18 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: After reading up on the Scala `Option` type, yes, using `Option[String]` would be helpful in handling the `None` case.  I found some `getOrElse` usages in other test code to mirror.

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 19 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: From what I understand, it's generic class that can be used for dynamic JSON navigation (i.e. the value might be a JSON string, array, number, object, etc)

Comment (markusthoemmes) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Right, but in this specific case, can we narrow it down to a more typed approach and thus easier to understand and read

File: tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala (hunk 4 above)
Comment (mdeuser) [inline tests/src/whisk/core/apigw/actions/test/ApiGwRoutemgmtActionTests.scala]: Good tips.  wsk.action.invoke is used in all of the the new positive test cases (with a properties file using the whisk.system credentials).

File: ansible/templates/whisk.properties.j2 [hunk 20]
```diff
@@ -86,3 +86,8 @@ db.password={{ db_password }}
 db.prefix={{ db_prefix }}
 db.whisk.actions={{ db_prefix }}whisks
 db.whisk.auths={{ db_auth }}
+db.whisk.apigw={{ db_apigw }}
+
+apigw.auth.user={{apigw_auth_user}}
+apigw.auth.pwd={{apigw_auth_pwd}}
+apigw.host={{apigw_host}}
```
Comment (domdom82) [inline ansible/templates/whisk.properties.j2]: why not change this to `http://{{ groups['apigateway']|first }}:{{apigateway.port.api}}/v1`
and remove from environment group vars? much cleaner this way and less overhead.

File: ansible/environments/local/group_vars/all [hunk 21]
```diff
@@ -26,3 +26,9 @@ limits:
   triggers:
     fires:
       perMinute: 60
+
+# API GW connection configuration
+apigw_auth_user: ""
+apigw_auth_pwd: ""
+apigw_host: "http://{{ groups['apigateway']|first }}:{{apigateway.port.api}}/v1"
```
Comment (mdeuser) [inline ansible/environments/local/group_vars/all]: @domdom82 - Yes, this will vary by environment.  In the OW environments it will always refer to a local API GW, in BW other fixed API GWs will be available.

**Comment by mdeuser** [comment]
@domdom82 - Recent commit has the ansible changes we talked about.  Thanks for the input!

**Comment by rabbah** [comment]
Keep in mind that `whisk.invoke` is now deprecated as of https://github.com/openwhisk/openwhisk/commit/85070a20f560e77ea064ebfc57f2c1eb4608f420.

File: core/routemgmt/activateApi.js [hunk 22]
```diff
@@ -0,0 +1,387 @@
...
+  return '';
+}
+
+function confidentialPrint(str) {
+    var printStr;
+    if (str) {
+        printStr = 'XXXXXXXXXX';
+    }
+    return printStr;
+}
```
Comment (markusthoemmes) [inline core/routemgmt/activateApi.js]: What is this for? Why not simply omit the print of credentials at all?

Comment (rabbah) [inline core/routemgmt/activateApi.js]: Let's please get all the credential logging out.

Comment (mdeuser) [inline core/routemgmt/activateApi.js]: if the credential is not supplied (indicates a corrupt package), the log will show an empty string for the credential.  non-empty credentials are not logged.

**Comment by mdeuser** [comment]
PG3 27 passed

**Comment by mdeuser** [comment]
@markusthoemmes - any other comments?  If you agree, any outstanding nice-to-have comments will have a separate issue opened up for them.

**Comment by mdeuser** [comment]
Travis CI build success.  PG1 972 and PG2 692 are both in progress

**Comment by csantanapr** [comment]
merged 06e8a075ca216fe37e21b3b33c50f7271e550455
